java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -178,190 +232,129 @@ final boolean isSuperOrThisMemberSelect(ExpressionTree tree) {
-        switch (currentState.evaluateRelation(leftValue, operator, rightValue)) {
-          case FALSE:
-            currentResult.falseStates.add(currentState);
-            break;
-          case TRUE:
-            currentResult.trueStates.add(currentState);
-            break;
-          default:
-            currentResult.falseStates.add(new ExecutionState(currentState).setRelation(leftValue, operator.negate(), rightValue));
-            currentResult.trueStates.add(new ExecutionState(currentState).setRelation(leftValue, operator, rightValue));
+    public final void visitConditionalExpression(ConditionalExpressionTree tree) {
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+      for (ExecutionState executionState : currentStates) {
+        for (ExecutionState conditionState : evaluateCondition(executionState, tree.condition())) {
+          SymbolicValue result = conditionState.pop();
+          if (conditionState.getBooleanConstraint(result) == SymbolicBooleanConstraint.FALSE) {
+            nextStates.addAll(evaluateWithState(conditionState, tree.falseExpression()));
+          } else {
+            nextStates.addAll(evaluateWithState(conditionState, tree.trueExpression()));
+          }
         }
-      } else {
-        currentResult.unknownStates.add(currentState);
       }
-    }
-  }
-
-  public class ExpressionVisitor extends BaseExpressionVisitor {
-    ExecutionState currentState;
-    SymbolicBooleanConstraint currentResult;
-
-    public SymbolicBooleanConstraint evaluate(ExecutionState state, ExpressionTree tree) {
-      currentState = state;
-      currentResult = SymbolicBooleanConstraint.UNKNOWN;
-      scan(tree);
-      return currentResult;
-    }
-
-    @Override
-    public final void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
-      super.visitArrayAccessExpression(tree);
-      currentResult = SymbolicBooleanConstraint.UNKNOWN;
+      currentStates = nextStates;
     }
 
     @Override
-    public final void visitAssignmentExpression(AssignmentExpressionTree tree) {
-      super.visitAssignmentExpression(tree);
-      Symbol.VariableSymbol symbol = extractVariableSymbol(tree.variable());
+    public final void visitIdentifier(IdentifierTree tree) {
+      Symbol.VariableSymbol symbol = extractVariableSymbol(tree);
       if (symbol != null) {
-        SymbolicValue.SymbolicVariableValue variable = new SymbolicValue.SymbolicVariableValue(symbol);
-        currentState.invalidateRelationsOnValue(variable);
-        currentState.setBooleanConstraint(variable, currentResult);
+        PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+        for (ExecutionState executionState : currentStates) {
+          nextStates.addState(executionState.push(executionState.getValue(symbol)));
+        }
+        currentStates = nextStates;
+      } else {
+        SymbolicValue value = new SymbolicValue();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this instantiation of "SymbolicValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)